{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HIV Analysis\n",
    "\n",
    "Here we combine HIV sequence data from the [Los Alamos National Laboratory HIV Sequence Database](www.hiv.lanl.gov) and immunological data to investigate HIV evolution across 14 individuals. This data is contained in the `data/HIV/` directory. Supplementary scripts for data processing are stored in `HIV.py`.\n",
    "\n",
    "We downloaded sequence data for all patients and aligned them with the [HXB2 reference sequence](https://www.hiv.lanl.gov/components/sequence/HIV/asearch/query_one.comp?se_id=K03455) and the corresponding [clade consensus sequence](https://www.hiv.lanl.gov/content/sequence/NEWALIGN/align.html#consensus) using the LANL tool [HIValign](https://www.hiv.lanl.gov/content/sequence/VIRALIGN/viralign.html). We recorded immunological data for these patients in `.csv` files for analysis.\n",
    "\n",
    "Additonal note (August.23.2023) How to get HIValign.fasta files. \n",
    "The steps to generate the files are just \n",
    "1. download all relevant sequences from LANL (xxx_gap_squeeze.fasta). \n",
    "2. Add the consensus sequence to the file (xxx_gap_squeeze_cons.fasta). \n",
    "3. Align the sequences together with the consensus sequence (xxx-HIValign.fasta). \n",
    "4. It looks like there is some subsequent processing here to remove gaps, but whatever this is, it should be in the notebook â€” this was not done by any other script.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### Sequence processing\n",
    "\n",
    "In the scripts below, we first process the HIV sequence data by\n",
    "1. associating times with sequences and removing all sequences for which no collection time is available,\n",
    "2. clipping full-length sequences down to the heavily-sequenced half-genome regions,\n",
    "3. removing sequences with excess gaps in the targeted region, which primarily consist of sequences that cover the opposite half-genome region,\n",
    "4. removing sites that are >= 95% gaps,\n",
    "5. time-ordering sequences,\n",
    "6. removing time points that are > 300 days from the previous time point and/or have < 4 sequences available at that time point, in order to ensure minimum quality of the correlations,\n",
    "6. imputing ambiguous nucleotides with the most frequently observed nucleotide at the same site (and optionally, imputing gaps at the sequence edges due to different sequencing windows; with or without imputation, gaps at the edges will be flagged), and\n",
    "7. saving the data in a format readable by the MPL program.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load libraries and define global variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This notebook was prepared using:\n",
      "python version 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 17:29:18) [GCC 11.2.0]\n",
      "numpy version 1.26.4\n",
      "pandas version 2.2.1\n"
     ]
    }
   ],
   "source": [
    "print('This notebook was prepared using:')\n",
    "\n",
    "import os\n",
    "import sys\n",
    "print('python version %s' % sys.version)\n",
    "\n",
    "import numpy as np\n",
    "print('numpy version %s' % np.__version__)\n",
    "\n",
    "import pandas as pd\n",
    "print('pandas version %s' % pd.__version__)\n",
    "\n",
    "import HIV\n",
    "\n",
    "# GLOBAL VARIABLES\n",
    "\n",
    "NUC = ['-', 'A', 'C', 'G', 'T']\n",
    "PRO = ['-', 'A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H',\n",
    "       'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V']\n",
    "REF = NUC[0]\n",
    "CONS_TAG = 'CONSENSUS'\n",
    "HXB2_TAG = 'B.FR.1983.HXB2-LAI-IIIB-BRU.K03455.19535'\n",
    "C_PPTS = ['703010505', '703010848']\n",
    "#C_PPTS = ['706010164']\n",
    "# # Code Ocean directories\n",
    "# HIV_DIR = '../data/HIV'\n",
    "# MPL_DIR = 'MPL'\n",
    "# HIV_MPL_DIR = '../data/HIV/MPL'\n",
    "\n",
    "# GitHub directories\n",
    "HIV_DIR = 'data/HIV'\n",
    "MPL_DIR = 'src/MPL'\n",
    "HIV_MPL_DIR = 'src/MPL/HIV'\n",
    "\n",
    "# Sequence processing\n",
    "MAX_GAP_NUM  = 200\n",
    "MAX_GAP_FREQ = 0.95\n",
    "MIN_SEQS     = 4\n",
    "MAX_DT       = 300"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/HIV/raw/703010848-HIValign.fasta\n",
      "data/HIV/raw/703010848-accession2time.dat\n",
      "data/HIV/raw/703010505-HIValign.fasta\n",
      "data/HIV/raw/703010505-accession2time.dat\n",
      "703010505-3\n",
      "\tselected 602 of 623 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 239 of 2891 sites with >95% gaps\n",
      "\texchanged C for Y in sequence 70, site 122\n",
      "\texchanged C for M in sequence 70, site 157\n",
      "\texchanged C for M in sequence 58, site 875\n",
      "\texchanged C for M in sequence 77, site 875\n",
      "\texchanged C for Y in sequence 170, site 935\n",
      "\texchanged A for R in sequence 100, site 954\n",
      "\texchanged C for Y in sequence 220, site 1029\n",
      "\texchanged A for R in sequence 112, site 1272\n",
      "\texchanged A for R in sequence 117, site 1272\n",
      "\texchanged C for Y in sequence 58, site 1273\n",
      "\texchanged C for Y in sequence 112, site 1273\n",
      "\texchanged C for Y in sequence 117, site 1273\n",
      "\texchanged C for M in sequence 220, site 1285\n",
      "\texchanged A for M in sequence 220, site 1287\n",
      "\texchanged A for R in sequence 170, site 1291\n",
      "\texchanged C for Y in sequence 102, site 1463\n",
      "\texchanged A for R in sequence 185, site 1678\n",
      "\texchanged A for R in sequence 254, site 2281\n",
      "\texchanged G for R in sequence 102, site 2564\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "703010848-3\n",
      "\tselected 1205 of 1215 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 158 of 4953 sites with >95% gaps\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\n",
      "Out of 24000 possible mutations,\n",
      "\t0 (0.00%) nonsynonymous in epitope\n",
      "\t516 (2.15%) nonsynonymous reversion\n",
      "\t0 (0.00%) nonsynonymous reversion, in epitope\n",
      "\t516 (2.15%) nonsynonymous reversion, nonepitope\n",
      "\t4565 (19.02%) synonymous\n"
     ]
    }
   ],
   "source": [
    "# Load the sequence coverage for each patient\n",
    "df_range = pd.read_csv('%s/range.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "\n",
    "# Load the list of transmitted/founder (TF) sequences\n",
    "df_TF = pd.read_csv('%s/TF.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "\n",
    "# Load the list of exposed residues on Env\n",
    "df_exposed = pd.read_csv('%s/exposed.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "\n",
    "# Load the set of epitopes targeted by patients\n",
    "df_epitope = pd.read_csv('%s/epitopes.csv' % (HIV_DIR), comment='#', dtype={'ppt': str}, memory_map=True)\n",
    "\n",
    "\n",
    "# Association times with sequences, remove all sequences (except HXB2 and consensus)\n",
    "# that do not have identifiable times; TF sequence is treated as day 0 if available\n",
    "\n",
    "for ppt in set(df_range.ppt):\n",
    "    HIV.filter_sequences(ppt, '%s/raw/%s-HIValign.fasta' % (HIV_DIR, ppt), check_time=True)\n",
    "    \n",
    "    \n",
    "# Store baseline number of nonsynonymous mutations in CD8+ T cell epitopes, \n",
    "# nonsynonymous reversions (total, in epitopes, and outside epitopes), and synonymous mutations\n",
    "BASELINE_NS_EPITOPE = 0\n",
    "BASELINE_NS_REV = 0\n",
    "BASELINE_NS_REV_NONEPI = 0\n",
    "BASELINE_SYN = 0\n",
    "BASELINE_TOT = 0\n",
    "\n",
    "# Select and process alignment subsets for analysis\n",
    "g = open('%s/run.sh' % (MPL_DIR), 'w')\n",
    "g.write('g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl\\n')\n",
    "\n",
    "for df_range_iter, df_range_entry in df_range.iterrows():\n",
    "    \n",
    "#     # Filter by subtype\n",
    "#     if df_range_entry.ppt not in C_PPTS:\n",
    "#         continue\n",
    "    \n",
    "    # Read in MSA\n",
    "    ppt = df_range_entry.ppt\n",
    "    msa, tag = HIV.get_MSA('%s/interim/%s-HIValign-filtered.fasta' % (HIV_DIR, ppt), noArrow=True)\n",
    "    print('%s' % df_range_entry.tag)\n",
    "    \n",
    "    # Clip MSA to specified range\n",
    "    msa = HIV.clip_MSA(df_range_entry.start, df_range_entry.end, msa, tag)\n",
    "    \n",
    "    # Filter sequences for quality (maximum #excess gaps)\n",
    "    current_msa, current_tag = HIV.filter_excess_gaps(msa, tag, MAX_GAP_NUM, MAX_GAP_FREQ, verbose=True)\n",
    "    \n",
    "    # Put sequences in time order (first entries are HXB2, consensus)\n",
    "    current_msa, current_tag = HIV.order_sequences(current_msa, current_tag)\n",
    "    \n",
    "    # Locate or compute TF sequence\n",
    "    TF_accession = df_TF[df_TF.ppt==ppt].iloc[0].TF\n",
    "    TF_sequence = HIV.get_TF(current_msa, current_tag, TF_accession)\n",
    "    \n",
    "    temp_msa = [current_msa[list(current_tag).index(HXB2_TAG)], TF_sequence]\n",
    "    temp_tag = [HXB2_TAG, '%s-TF' % ppt]\n",
    "    HIV.save_MSA(temp_msa, temp_tag, '%s/interim/TF-%s' % (HIV_DIR, df_range_entry.tag))\n",
    "    \n",
    "    # Impute ambiguous nucleotides and gaps at start/end of alignment\n",
    "    current_msa = HIV.impute_ambiguous(current_msa, current_tag, start_index=2, verbose=True, impute_edge_gaps=False)\n",
    "    \n",
    "    # Map between local (alignment) index, index of polymorphic sites, and HXB2\n",
    "    # Track location of known T cell epitopes, TF sequence, and exposed residues of Env\n",
    "    poly_sites, current_msa, current_tag = HIV.create_index(current_msa, current_tag, TF_sequence, \n",
    "                                           current_msa[list(current_tag).index(CONS_TAG)], \n",
    "                                           current_msa[list(current_tag).index(HXB2_TAG)],\n",
    "                                           df_range_entry.start, MIN_SEQS, MAX_DT, \n",
    "                                           df_epitope[df_epitope.ppt==ppt], df_exposed, \n",
    "                                           '%s/processed/%s-index.csv' % (HIV_DIR, df_range_entry.tag), \n",
    "                                           return_polymorphic=True, return_truncated=True)\n",
    "    \n",
    "    # Save alignment to format readable by MPL\n",
    "    poly_states, poly_times = HIV.save_MPL_alignment(current_msa, current_tag, \n",
    "                                                     '%s/%s-poly-seq2state.dat' % (HIV_MPL_DIR, df_range_entry.tag), \n",
    "                                                     polymorphic_sites=poly_sites, return_states=True)\n",
    "    \n",
    "    # Save allele frequency trajectories for polymorphic sites\n",
    "    # Later this file will be edited to include inferred selection coefficients\n",
    "    df_index = pd.read_csv('%s/processed/%s-index.csv' % (HIV_DIR, df_range_entry.tag), comment='#', memory_map=True)\n",
    "    HIV.save_trajectories(poly_sites, poly_states, poly_times, TF_sequence, df_index,\n",
    "                          '%s/interim/%s-poly.csv' % (HIV_DIR, df_range_entry.tag))\n",
    "    HIV.save_trajectories_alternate(poly_sites, poly_states, poly_times, TF_sequence, df_index,\n",
    "                          '%s/interim/%s-poly-alternate.csv' % (HIV_DIR, df_range_entry.tag))\n",
    "    \n",
    "    # Get baseline number of possible mutations in different categories\n",
    "    if ppt not in ['cap256']:\n",
    "        df_poly = pd.read_csv('%s/interim/%s-poly.csv' % (HIV_DIR, df_range_entry.tag), comment='#', memory_map=True)\n",
    "        temp_ns_epitope, temp_ns_rev, temp_ns_rev_ne, temp_syn = HIV.get_baseline(TF_sequence, df_index, df_poly)\n",
    "        BASELINE_NS_EPITOPE += temp_ns_epitope\n",
    "        BASELINE_NS_REV += temp_ns_rev\n",
    "        BASELINE_NS_REV_NONEPI += temp_ns_rev_ne\n",
    "        BASELINE_SYN += temp_syn\n",
    "        BASELINE_TOT += 3 * len(df_index) + np.sum(df_poly.nucleotide=='-')\n",
    "    \n",
    "    # Save MSA\n",
    "    HIV.save_MSA(current_msa, current_tag, '%s/interim/%s' % (HIV_DIR, df_range_entry.tag))\n",
    "    \n",
    "    # Record instructions for running MPL on this data\n",
    "    io_str = '%s-poly-seq2state' % df_range_entry.tag\n",
    "    run_str = ('./bin/mpl -d HIV -i %s.dat -o %s-MPL.dat -g 5e4 -N 1e4 -m Zanini-extended.dat -sc %s -sn %s\\n'+\n",
    "               './bin/mpl -d HIV -i %s.dat -o %s-SL.dat  -g 5e4 -N 1e4 -m Zanini-extended.dat -nc\\n')\n",
    "    g.write(run_str % (io_str, io_str, 'covariance-'+io_str+'.dat', 'numerator-'+io_str+'.dat', io_str, io_str))\n",
    "\n",
    "g.close()\n",
    "\n",
    "\n",
    "# Print baseline numbers of possible mutations in different categories\n",
    "print('\\nOut of %d possible mutations,' % BASELINE_TOT)\n",
    "print('\\t%d (%.2f%%) nonsynonymous in epitope' \n",
    "      % (BASELINE_NS_EPITOPE, 100*BASELINE_NS_EPITOPE/BASELINE_TOT))\n",
    "print('\\t%d (%.2f%%) nonsynonymous reversion'\n",
    "      % (BASELINE_NS_REV, 100*BASELINE_NS_REV/BASELINE_TOT))\n",
    "print('\\t%d (%.2f%%) nonsynonymous reversion, in epitope' \n",
    "      % (BASELINE_NS_REV-BASELINE_NS_REV_NONEPI, 100*(BASELINE_NS_REV-BASELINE_NS_REV_NONEPI)/BASELINE_TOT))\n",
    "print('\\t%d (%.2f%%) nonsynonymous reversion, nonepitope' \n",
    "      % (BASELINE_NS_REV_NONEPI, 100*BASELINE_NS_REV_NONEPI/BASELINE_TOT))\n",
    "print('\\t%d (%.2f%%) synonymous'\n",
    "      % (BASELINE_SYN, 100*BASELINE_SYN/BASELINE_TOT))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
